cpp/include/cudf/aggregation.hpp

```diff
@@ -82,108 +84,136 @@ class aggregation {
     CUDA             ///< CUDA UDF based reduction
   };
 
+  aggregation() = delete;
   aggregation(aggregation::Kind a) : kind{a} {}
   Kind kind;  ///< The aggregation to perform
+  virtual ~aggregation() = default;
 
   virtual bool is_equal(aggregation const& other) const { return kind == other.kind; }
-
   virtual size_t do_hash() const { return std::hash<int>{}(kind); }
+  virtual std::unique_ptr<aggregation> clone() const = 0;
 
-  virtual std::unique_ptr<aggregation> clone() const
-  {
-    return std::make_unique<aggregation>(*this);
-  }
+  // override functions for compound aggregations
+  virtual std::vector<std::unique_ptr<aggregation>> get_simple_aggregations(
+    data_type col_type, cudf::detail::simple_aggregations_collector& collector) const = 0;
+  virtual void finalize(cudf::detail::aggregation_finalizer& finalizer)               = 0;
+};
 
-  virtual ~aggregation() = default;
+/**
+ * @brief Derived class intended for enforcing operation-specific restrictions
+ * when calling various cudf functions.
+ *
+ * As an example, rolling_window will only accept rolling_aggregation inputs,
+ * and the appropriate derived classes (sum_aggregation, mean_aggregation, etc)
+ * derive from this interface to represent these valid options.
+ */
+class rolling_aggregation : public virtual aggregation {
+ public:
+  ~rolling_aggregation() = default;
 
-  // override functions for compound aggregations
-  virtual std::vector<aggregation::Kind> get_simple_aggregations(data_type col_type) const;
-  virtual void finalize(cudf::detail::aggregation_finalizer& finalizer);
+ protected:
+  rolling_aggregation(){};
 };
 
 enum class udf_type : bool { CUDA, PTX };
 
 /// Factory to create a SUM aggregation
-std::unique_ptr<aggregation> make_sum_aggregation();
+template <typename Base = aggregation>
+extern std::unique_ptr<Base> make_sum_aggregation();
```

Comment:
I don't think these `extern`s are necessary. Normally you `extern` a particular instantiation of the template to tell the linker to go look for that code somewhere else at link time instead of reinstantiating the template with the available definition.

Here, there is no available definition, so it has to go look for an existing instantiation and will get an undefined symbol error if it doesn't exist.